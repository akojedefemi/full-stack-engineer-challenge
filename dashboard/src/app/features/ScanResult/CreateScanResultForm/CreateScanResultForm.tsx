import React from 'react';
import {
  Form,
  Input,
  Tooltip,
  Icon,
  Cascader,
  Select,
  Row,
  Col,
  Checkbox,
  Button,
  AutoComplete,
} from 'antd';
import { FormComponentProps } from 'antd/lib/form';

import { ScanResultFormData } from '../models/ScanResultFormData';

const formItemLayout = {
  labelCol: {
    xs: { span: 12 },
    sm: { span: 4 },
  },
  wrapperCol: {
    xs: { span: 12 },
    sm: { span: 8 },
  },
};

const hasErrors = (fieldsError: Record<string, string[] | undefined>) => {
  return Object.keys(fieldsError).some(field => fieldsError[field]);
}

const validateJson = (rule: any, value: any) => {
  try {
    const result = JSON.parse(value);
    return !!result;
  } catch (error) {
    return false;
  }
}

interface Props extends FormComponentProps<ScanResultFormData> {
  onSubmit: (values: ScanResultFormData) => void;
}

const CreateScanResultFormComponent: React.FC<Props> = (props) => {
  const { form, onSubmit } = props;
  const { getFieldDecorator, getFieldsError, validateFields } = form;

  const handleSubmit = (event: React.FormEvent<HTMLFormElement>) => {
    event.preventDefault();

    validateFields((error, values) => {
      if (!error) {
        onSubmit(values);
      }
    });
  };

  const renderSubmitButton = () => {
    return (
      <Form.Item>
        <Button type="primary" htmlType="submit" disabled={hasErrors(getFieldsError())}>
          Save
        </Button>
      </Form.Item>
    )
  }

  const renderNameInput = () => {
    return (
      <Form.Item label="Repository Name">
        {getFieldDecorator('repositoryName', {
          rules: [{ required: true, message: 'Please input the name!', whitespace: true }],
        })(<Input />)}
      </Form.Item>
    )
  }

  const renderFindingsInput = () => {
    return (
      <Form.Item label="Findings">
        {getFieldDecorator('findings', {
          rules: [{ required: true, message: 'Please paste findings Data!', whitespace: true, validator: validateJson, }],
        })(<Input />)}
      </Form.Item>
    )
  }

  return (
    <Form {...formItemLayout} onSubmit={handleSubmit}>
      {renderNameInput()}
      {renderFindingsInput()}
      {renderSubmitButton()}
    </Form>
  )
};

const CreateScanResultForm = Form.create<Props>({ name: 'Create Scan Result' })(CreateScanResultFormComponent);

export default CreateScanResultForm;
