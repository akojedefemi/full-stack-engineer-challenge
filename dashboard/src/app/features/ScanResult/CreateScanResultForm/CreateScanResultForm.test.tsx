import React from 'react';
import renderer from 'react-test-renderer';
import { BrowserRouter } from 'react-router-dom';

import CreateScanResultForm from './CreateScanResultForm';

const noop = () => { };

describe('CreateScanResultForm', () => {

  it('renders correctly', () => {
    expect(renderer.create(<BrowserRouter><CreateScanResultForm onSubmit={noop} /></BrowserRouter>).toJSON()).toMatchSnapshot();
  });

});
