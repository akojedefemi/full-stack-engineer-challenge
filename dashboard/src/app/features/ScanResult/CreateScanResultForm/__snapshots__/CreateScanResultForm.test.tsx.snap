// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreateScanResultForm renders correctly 1`] = `
<form
  className="ant-form ant-form-horizontal"
  onSubmit={[Function]}
>
  <div
    className="ant-row ant-form-item"
    style={Object {}}
  >
    <div
      className="ant-col ant-form-item-label ant-col-xs-12 ant-col-sm-4"
      style={Object {}}
    >
      <label
        className="ant-form-item-required"
        htmlFor="Create Scan Result_repositoryName"
        onClick={[Function]}
        title="Repository Name"
      >
        Repository Name
      </label>
    </div>
    <div
      className="ant-col ant-form-item-control-wrapper ant-col-xs-12 ant-col-sm-8"
      style={Object {}}
    >
      <div
        className="ant-form-item-control"
      >
        <span
          className="ant-form-item-children"
        >
          <input
            className="ant-input"
            data-__field={
              Object {
                "name": "repositoryName",
              }
            }
            data-__meta={
              Object {
                "name": "repositoryName",
                "originalProps": Object {
                  "type": "text",
                },
                "ref": null,
                "rules": Array [
                  Object {
                    "message": "Please input the name!",
                    "required": true,
                    "whitespace": true,
                  },
                ],
                "trigger": "onChange",
                "validate": Array [
                  Object {
                    "rules": Array [
                      Object {
                        "message": "Please input the name!",
                        "required": true,
                        "whitespace": true,
                      },
                    ],
                    "trigger": Array [
                      "onChange",
                    ],
                  },
                ],
                "valuePropName": "value",
              }
            }
            id="Create Scan Result_repositoryName"
            onChange={[Function]}
            onKeyDown={[Function]}
            type="text"
            value=""
          />
        </span>
      </div>
    </div>
  </div>
  <div
    className="ant-row ant-form-item"
    style={Object {}}
  >
    <div
      className="ant-col ant-form-item-label ant-col-xs-12 ant-col-sm-4"
      style={Object {}}
    >
      <label
        className="ant-form-item-required"
        htmlFor="Create Scan Result_findings"
        onClick={[Function]}
        title="Findings"
      >
        Findings
      </label>
    </div>
    <div
      className="ant-col ant-form-item-control-wrapper ant-col-xs-12 ant-col-sm-8"
      style={Object {}}
    >
      <div
        className="ant-form-item-control"
      >
        <span
          className="ant-form-item-children"
        >
          <input
            className="ant-input"
            data-__field={
              Object {
                "name": "findings",
              }
            }
            data-__meta={
              Object {
                "name": "findings",
                "originalProps": Object {
                  "type": "text",
                },
                "ref": null,
                "rules": Array [
                  Object {
                    "message": "Please paste findings Data!",
                    "required": true,
                    "validator": [Function],
                    "whitespace": true,
                  },
                ],
                "trigger": "onChange",
                "validate": Array [
                  Object {
                    "rules": Array [
                      Object {
                        "message": "Please paste findings Data!",
                        "required": true,
                        "validator": [Function],
                        "whitespace": true,
                      },
                    ],
                    "trigger": Array [
                      "onChange",
                    ],
                  },
                ],
                "valuePropName": "value",
              }
            }
            id="Create Scan Result_findings"
            onChange={[Function]}
            onKeyDown={[Function]}
            type="text"
            value=""
          />
        </span>
      </div>
    </div>
  </div>
  <div
    className="ant-row ant-form-item"
    style={Object {}}
  >
    <div
      className="ant-col ant-form-item-control-wrapper ant-col-xs-12 ant-col-sm-8"
      style={Object {}}
    >
      <div
        className="ant-form-item-control"
      >
        <span
          className="ant-form-item-children"
        >
          <button
            className="ant-btn ant-btn-primary"
            disabled={false}
            onClick={[Function]}
            type="submit"
          >
            <span>
              Save
            </span>
          </button>
        </span>
      </div>
    </div>
  </div>
</form>
`;
